readthedocs/vcs_support/backends/git.py

```diff
@@ -68,6 +91,166 @@ def update(self):
+
+        # TODO: This seems to be legacy that can be removed.
+        #  There shouldn't be cases where we are asked to
+        #  clone the repo in a non-clean working directory.
+        #  The prior call to repo_exists() will return if a repo already exist with
+        #  unclear guarantees about whether that even needs to be a fully consistent clone.
+        self.make_clean_working_dir()
+
+        # --no-checkout: Makes it explicit what we are doing here. Nothing is checked out
+        #                until it's explicitly done.
+        # --depth 1: Shallow clone, fetch as little data as possible.
+        cmd = ["git", "clone", "--depth", "1"]
+
+        # We add "--no-checkout" in all git clone operations, except:
+        # There exists a case of version_type=BRANCH without a branch name.
+        # This case is relevant for building projects for the first time without knowing the name
+        # of the default branch.
+        if self.version_type == BRANCH and not self.version_identifier:
+            self._skip_fetch = True
+        else:
+            cmd.append("--no-checkout")
+
+        cmd += [self.repo_url, "."]
+
+        try:
+            # TODO: Explain or remove the return value
+            code, stdout, stderr = self.run(*cmd)
+            return code, stdout, stderr
+        except RepositoryError as exc:
+            raise RepositoryError(RepositoryError.CLONE_ERROR()) from exc
+
+    def fetch_ng(self):
+        """Implementation for new clone+fetch+checkout pattern."""
+
+        # When git clone does NOT add --no-checkout, it's because we are going
+        # to use the remote HEAD, so we don't have to fetch nor check out.
+        if self._skip_fetch:
+            log.info("Skipping git fetch")
+            return
+
+        # --force: Likely legacy, it seems to be irrelevant to this usage
+        # --prune: Likely legacy, we don't expect a previous fetch command to have run
+        # --prune-tags: Likely legacy, we don't expect a previous fetch command to have run
+        # --tags: This flag was used in the previous approach such that all tags were fetched
+        #         in order to checkout a tag afterwards.
+        # --depth: This flag should be made unnecessary, it's downloading commit data that's
+        #          never used.
+        cmd = [
+            "git",
+            "fetch",
+            "origin",
+            "--force",
+            "--prune",
+            "--prune-tags",
+            "--depth",
+            str(self.repo_depth),
+        ]
+        remote_reference = self.get_remote_fetch_refspec()
+
+        if remote_reference:
+            # TODO: We are still fetching the latest 50 commits.
+            # A PR might have another commit added after the build has started...
+            cmd.append(remote_reference)
+
+        # Log a warning, except for machine versions since it's a known bug that
+        # we haven't stored a remote refspec in Version for those "stable" versions.
```

Comment:
This only happens for the first build of the project, right? When we don't yet know the identifier? This also happens to `latest` version, right?

Reply:
The `latest` version has the default branch indicated as `version_identifier`.

Reply:
> The latest version has the default branch indicated as version_identifier.

I think this is only true _after_ the initial version synchronization is done. The "default branch" could be `None` when the project is imported so we don't know what is its name yet and we need to perform a clone without a specific fetch/checkout.

Reply:
Ah yes, but this is what the warning below is warning about :) I've updated the comment.